### PR TITLE
fix(common)!: Validate processor affinity against mask width only

### LINKED ISSUE #257: ProcessExtensions: Environment.ProcessorCount is not a valid affinity-bit index cap for non-contiguous affinities
## Background

Follow-up from PR #256 review (CodeRabbit finding: https://github.com/mrploch/ploch-common/pull/256#discussion_r3646854677).

Since .NET 6, `Environment.ProcessorCount` returns the number of CPUs **available to the process** — the minimum of the machine's logical CPU count, the process's own affinity CPU count, and any CPU-utilisation limit. It is therefore a *count*, not an upper bound on valid CPU *indices*.

## Problem

`ProcessExtensions` (both before and after #249/#256) uses `Environment.ProcessorCount` as a bit-index cap:

- `GetEnabledProcessors` loops `i < Math.Min(Environment.ProcessorCount, IntPtr.Size * 8)`. For a process already constrained to a non-contiguous CPU set such as 8–15 (`ProcessorCount == 8`), it only inspects bits 0–7 and reports **no** enabled processors.
- `ValidateProcessorNumber` rejects CPU numbers ≥ `ProcessorCount` even when they are valid indices within the native mask (e.g. CPU 12 in the 8–15 example).

This is pre-existing behaviour on `master`; #256 only added the (correct) native-mask-width cap on top of it.

## Suggested direction

- Bound `GetEnabledProcessors` by the native mask width alone — reading high bits is harmless and correctly reports non-contiguous affinity sets.
- Decide the validation contract for the setters: keep rejecting indices ≥ `ProcessorCount` (simple, but wrong for constrained processes), or validate only against the mask width and let the OS reject genuinely non-existent CPUs. This changes documented behaviour (`ArgumentOutOfRangeException` message and XML docs updated in #256), so it needs a deliberate design decision and tests for constrained-affinity scenarios.

Refs: #249, PR #256

### PR DESCRIPTION
**Pull request by kploch**
### Summary

Implements #257: `Environment.ProcessorCount` is no longer used as an upper bound for processor numbers in `Ploch.Common.Diagnostics.ProcessExtensions`. Since .NET 6, `ProcessorCount` is the number of processors *available to the process* (it already reflects the process's own affinity and CPU limits) — it is a count, not an index bound. Using it as an affinity-bit index cap rejected valid processor numbers and under-reported enabled processors for processes constrained to non-contiguous processor sets (e.g. CPUs 8–15, where `ProcessorCount == 8`).

### Changes

- `ValidateProcessorNumber` now validates only against the native affinity-mask width (`IntPtr.Size * 8` — 32 in a 32-bit process, 64 in a 64-bit process); the `processorCount` parameter is removed (`src/Common/Diagnostics/ProcessExtensions.cs`).
- `GetEnabledProcessors` reports every bit set in the native mask via a new internal, unit-testable core `GetEnabledProcessors(long affinityMask, int affinityMaskWidth)`, so non-contiguous affinity sets are reported correctly.
- `MaxAddressableProcessors` and `GetMaxAddressableProcessors` (internal helpers introduced in #256) are removed — the mask width is the single bound.
- XML documentation updated on all three public methods: processor numbers within the mask width are not validated against `ProcessorCount`; CPUs that do not exist on the machine are rejected by the operating system when the affinity is applied (typically `Win32Exception`).
- Change-log entry added (`change-log/2026-07-24-257-affinity-mask-width-only-validation.md`).

### Design decisions

- **Mask width as the sole eager bound.** The library validates what it can know reliably (representability in the native mask); actual CPU existence is the OS's call. Machine-wide logical CPU count is not reliably obtainable cross-platform without P/Invoke, and `ProcessorCount` is affinity-aware, so any eager count-based check would be wrong for constrained processes.
- **Internal `GetEnabledProcessors(long, int)` core** keeps the bit-extraction logic unit-testable for masks/widths that do not occur on the test machine (non-contiguous sets, 32-bit width), consistent with the `InternalsVisibleTo` approach from #256.

### Behaviour change (breaking)

Calls with a processor number within the native mask width but ≥ `Environment.ProcessorCount` previously threw `ArgumentOutOfRangeException` eagerly. They are now passed to the operating system: valid targets (processors the machine has, under a constrained affinity) succeed, and nonexistent processors surface as an OS-level error (typically `Win32Exception`) when the affinity is applied.

### Testing

- `ValidateProcessorNumber` theories rewritten for the new signature (throw: width 64/32 boundaries, 127, −1; accept: 63@64, 31@32, 0).
- New mask-extraction theories on the internal core: non-contiguous set `0xFF00 → [8..15]`, single/zero-bit masks, bit 32 excluded at width 32, and both 64-bit extremes (`[0, 63]`).
- The obsolete `ProcessorCount`-cap test is removed with the contract; existing child-process round-trip tests are unchanged.
- Full solution builds with zero new warnings; `Ploch.Common.Tests`: 852 passed (net10.0) / 819 passed (net8.0), 0 failed.

### Reviews performed before push

- Gemini code review: **APPROVED_WITH_NOTES** (notes: document the contract change — done in XML docs + change-log; platform caveats — already documented).
- Independent local review agent (substituting the Codex MCP, which rejects all models on this account): **APPROVED** — verified no residual `ProcessorCount` capping, no coverage holes, docs and change-log accuracy.

### Review-driven additions

- **Applied-affinity verification (commit 40777fc):** `SetEnabledProcessors` now reads the OS-reported mask back after applying it and throws `InvalidOperationException` listing any requested processors that were not enabled. Linux applies the *intersection* of the requested mask and the processors available to the process (`sched_setaffinity(2)`) rather than rejecting unavailable ones, so a mixed request (e.g. `{0, 63}` on a small host) would otherwise succeed only partially and silently (chatgpt-codex-connector P2 finding). `SetSingleProcessorAffinity` needs no verification — a single-CPU mask either applies fully or produces an empty intersection, which the OS rejects. The check is implemented as an internal, unit-tested core (`VerifyAppliedAffinity`), including a 32-bit sign-extension case.
- **Doc wording:** `AffinityMaskWidth` XML doc now says "representable" instead of "addressable" (Copilot suggestion).
- **DeepSource CS-P1009 (commit e281c0e):** empty-array `[InlineData]` rows replaced with dedicated `BeEmpty()` facts (`Array.Empty<int>()` is not usable in attribute arguments).

- **Later review rounds (commits 46f9bef → 6c594b4):** DeepSource CS-R1019 (index-from-end); `ValidateProcessorNumber` now guards its own mask-width parameter (1–64); the legacy Windows-only round-trip test picks its target from the child's OS-reported mask, and kills/reaps + disposes the child; `VerifyAppliedAffinity` lists missing bits over the full 64-bit width so its message is bitness-independent (CodeRabbit); Sourcery's zero-width-mask and single-bit-mask boundary tests added; exception message reworded to stay focused on representability (Copilot). One Copilot thread (null-check on `Process.Start(string)`) was resolved as a false positive with compiler evidence — that overload is annotated non-nullable.

## Issue ticket number and link

Closes #257 — https://github.com/mrploch/ploch-common/issues/257

## Checklist before requesting a review
- [x] I have performed a self-review of my code
- [x] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? — N/A
- [ ] Will this be part of a product update? If yes, please write one phrase about this update. — Yes: processor-affinity extensions now support non-contiguous constrained affinities; validation is against the native mask width, with nonexistent CPUs rejected by the OS (part of v4.0).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by Sourcery

Validate processor affinity solely against the native mask width and ensure applied affinity matches requested processors, improving correctness for constrained and non-contiguous CPU sets.

New Features:
- Add verification that SetEnabledProcessors successfully enables all requested processors and throw InvalidOperationException when the OS applies only a subset.

Bug Fixes:
- Stop using Environment.ProcessorCount as an upper bound for processor indices so affinity operations work correctly for processes constrained to non-contiguous or higher-indexed CPUs.

Enhancements:
- Report all enabled processors directly from the native affinity mask via a reusable GetEnabledProcessors core that is independent of Environment.ProcessorCount.
- Tighten processor-number validation around affinity-mask representability and make mask-width constraints explicit and OS-driven.

Documentation:
- Update XML documentation and add a change-log entry to describe the new affinity validation contract and behavior across platforms.

Tests:
- Expand and restructure ProcessExtensions tests to cover mask-width-only validation, non-contiguous affinity masks, applied-affinity verification, and edge cases such as 32-bit mask widths and invalid mask sizes.

## Summary by CodeRabbit

- **Breaking Changes**
  - Processor affinity validation now uses the supported affinity-mask width rather than the system’s processor count.
  - Requests for unavailable processors may now return operating-system errors instead of immediate validation errors.

- **Bug Fixes**
  - Affinity queries now correctly report non-contiguous enabled processors.
  - Affinity updates detect and report when requested processors were not applied.

### CODE REVIEW

File: src/Common/Diagnostics/ProcessExtensions.cs [hunk 1]
```diff
@@ -12,25 +12,20 @@ public static class ProcessExtensions
 {
     /// <summary>
     /// Gets the width, in bits, of the native <see cref="Process.ProcessorAffinity"/> mask for the current process
-    /// (32 in a 32-bit process, 64 in a 64-bit process).
+    /// (32 in a 32-bit process, 64 in a 64-bit process). This is the exclusive upper bound for addressable
+    /// processor numbers.
```
[the remark this answers is not in the document]
Comment (kploch) [reply src/Common/Diagnostics/ProcessExtensions.cs]: Done in 40777fc — the `AffinityMaskWidth` XML doc now describes the width as "the exclusive upper bound for processor numbers representable in the mask", matching `ValidateProcessorNumber` and the public methods' remarks. Thanks!

File: src/Common/Diagnostics/ProcessExtensions.cs [hunk 2]
```diff
@@ -91,7 +102,7 @@ public static void SetEnabledProcessors(this Process process, params int[] enabl
         long affinityMask = 0;
         foreach (var number in enabledProcessorsNumbers)
         {
-            ValidateProcessorNumber(number, Environment.ProcessorCount, AffinityMaskWidth, nameof(enabledProcessorsNumbers));
+            ValidateProcessorNumber(number, AffinityMaskWidth, nameof(enabledProcessorsNumbers));
```
[the remark this answers is not in the document]
Comment (kploch) [reply src/Common/Diagnostics/ProcessExtensions.cs:110]: Excellent catch — accepted and fixed in 40777fc. `SetEnabledProcessors` now reads the OS-reported mask back after applying it and throws `InvalidOperationException` listing every requested processor that was not enabled, so the Linux `sched_setaffinity(2)` intersection behaviour surfaces as an error instead of a silent partial success. The check is an internal, unit-tested core (`VerifyAppliedAffinity`, subset check `(applied & requested) == requested`), with theories covering the dropped-CPU-63 scenario you described, a dropped low bit, an applied-superset mask, and 32-bit `IntPtr.ToInt64()` sign-extension of bit 31. `SetSingleProcessorAffinity` deliberately has no verification: a single-CPU mask either applies fully or produces an empty intersection, which the OS already rejects (EINVAL → `Win32Exception`).

File: src/Common/Diagnostics/ProcessExtensions.cs [hunk 3]
```diff
@@ -122,12 +136,25 @@ public static IEnumerable<int> GetEnabledProcessors(this Process process)
...
+    /// </remarks>
+    internal static IEnumerable<int> GetEnabledProcessors(long affinityMask, int affinityMaskWidth)
+    {
         var enabledProcessors = new List<int>();
-        var maxAddressableProcessors = MaxAddressableProcessors;
-        for (var i = 0; i < maxAddressableProcessors; i++)
+        for (var i = 0; i < affinityMaskWidth; i++)
         {
             if ((affinityMask & (1L << i)) != 0)
             {
```
[the remark this answers is not in the document]
Comment (kploch) [reply src/Common/Diagnostics/ProcessExtensions.cs:183]: Done — the internal `GetEnabledProcessors(long, int)` core now validates `affinityMaskWidth` up front and throws `ArgumentOutOfRangeException` for widths a 64-bit mask cannot represent, making the contract explicit instead of silently wrapping the `1L << i` shift (commit 6fa8f71, with theories for widths 65 and −1). Implementation note: the guard uses a single unsigned comparison (`(uint)affinityMaskWidth > 64`) rather than a two-clause range check, because SonarCloud suggested a relational pattern (IDE0078) that is C# 9 while the netstandard2.0 target compiles with C# 7.3 (commit 3d8ac14). `VerifyAppliedAffinity` (your line-184 sibling) delegates to this same guarded core, so it is covered by the same check. Thanks!

**Comment by kploch** [comment]
@sourcery-ai review

File: src/Common/Diagnostics/ProcessExtensions.cs [hunk 4]
```diff
@@ -139,39 +187,51 @@ public static IEnumerable<int> GetEnabledProcessors(this Process process)
...
+    internal static void ValidateProcessorNumber(int processorNumber, int affinityMaskWidth, string paramName)
     {
-        var maxAddressable = GetMaxAddressableProcessors(processorCount, affinityMaskWidth);
-        if (processorNumber < 0 || processorNumber >= maxAddressable)
+        if (processorNumber < 0 || processorNumber >= affinityMaskWidth)
         {
             throw new ArgumentOutOfRangeException(paramName,
                                                   processorNumber,
-                                                  $"Processor number must be between 0 and {maxAddressable - 1} — the lesser of the machine's processor count ({processorCount}) and the native affinity-mask width ({affinityMaskWidth} bits).");
+                                                  $"Processor number must be between 0 and {affinityMaskWidth - 1} — the width of the native processor-affinity mask ({affinityMaskWidth} bits). Processor numbers for CPUs that do not exist on this machine are rejected by the operating system when the affinity is applied.");
```
[the remark this answers is not in the document]
Comment (kploch) [reply src/Common/Diagnostics/ProcessExtensions.cs]: Done in c701211 — `ValidateProcessorNumber` now guards `affinityMaskWidth` up front (valid range 1–64, single unsigned comparison consistent with the extraction core's guard) and throws `ArgumentOutOfRangeException` naming `affinityMaskWidth`, so an invalid width can neither produce the nonsensical "between 0 and -1" message nor admit processor numbers whose `1L << n` shift would wrap. Covered by new theories for widths 0, 65 and −1. Thanks!

File: tests/Common.Tests/Diagnostics/ProcessExtensionsTests.cs [hunk 5]
```diff
@@ -191,9 +179,8 @@ private static int[] GetAllowedProcessors(Process process)
     {
         // Derived from the raw OS-reported mask (not from GetEnabledProcessors) so it stays independent of the code under test.
         var affinityMask = process.ProcessorAffinity.ToInt64();
-        var maxAddressable = Math.Min(Environment.ProcessorCount, IntPtr.Size * 8);
 
-        return Enumerable.Range(0, maxAddressable).Where(processor => (affinityMask & (1L << processor)) != 0).ToArray();
+        return Enumerable.Range(0, IntPtr.Size * 8).Where(processor => (affinityMask & (1L << processor)) != 0).ToArray();
```
[the remark this answers is not in the document]
Comment (kploch) [reply tests/Common.Tests/Diagnostics/ProcessExtensionsTests.cs:191]: Done in c701211 — the legacy Windows-only round-trip test now picks its target from the child's OS-reported affinity mask (`GetAllowedProcessors(process)[^1]`) instead of `Environment.ProcessorCount - 1`, and asserts the initial enabled set against the OS-reported mask rather than a `ProcessorCount` count. As a bonus the child process is now killed and reaped in a `finally` instead of leaking. Thanks!

File: tests/Common.Tests/Diagnostics/ProcessExtensionsTests.cs [hunk 6]
```diff
@@ -13,14 +13,22 @@ public void SetSingleProcessorAffinity_should_set_affinity_mask_for_valid_proces
     {
         var process = Process.Start("../../../../../src/TestingSupport.MockConsoleApp/bin/Debug/net10.0/Ploch.TestingSupport.MockConsoleApp.exe");
```
[the remark this answers is not in the document]
Comment (kploch) [reply tests/Common.Tests/Diagnostics/ProcessExtensionsTests.cs]: I believe this is a false positive. The `Process.Start(string)` overload used here is annotated as returning non-nullable `Process` — it throws (e.g. `Win32Exception`) rather than returning null when the process cannot be started; only the `Process.Start(ProcessStartInfo)` overload is annotated `Process?` (it can return null when an existing process is reused). Two pieces of evidence: (1) this repository compiles with `Nullable=enable` and nullable warnings elevated to errors, and the unguarded dereference of the return value in this very test builds cleanly — if the overload returned `Process?`, CS8602 would fail the build; (2) a minimal probe project dereferencing `Process.Start("x")` under `#nullable enable` with `-warnaserror` compiles without warnings, while the `ProcessStartInfo` overload requires null-handling (which is why `StartShortLivedChildProcess` in this same file uses the null-forgiving operator on that overload). A failed start in this test surfaces as a thrown exception with full diagnostics, not a later `NullReferenceException`. Resolving as false positive.

File: tests/Common.Tests/Diagnostics/ProcessExtensionsTests.cs [hunk 7]
```diff
@@ -222,23 +218,77 @@ public void ValidateProcessorNumber_should_throw_when_processor_number_exceeds_a
...
 
         act.Should().NotThrow();
     }
 
     [Theory]
-    [InlineData(128, 64, 64)] // Machine with >64 logical CPUs, 64-bit process: only the first 64 CPUs are addressable.
-    [InlineData(64, 32, 32)] // 32-bit process on a 64-CPU machine: only the first 32 CPUs are addressable.
-    [InlineData(8, 64, 8)] // Typical machine: the processor count is the limiting factor.
-    public void GetMaxAddressableProcessors_should_cap_at_the_lesser_of_processor_count_and_mask_width(int processorCount, int affinityMaskWidth, int expected)
+    [InlineData(unchecked((long)0x8000000000000001UL), 0x1L, "63")] // Linux silently drops the unavailable CPU 63 (sched_setaffinity intersects).
```
[the remark this answers is not in the document]
Comment (kploch) [reply tests/Common.Tests/Diagnostics/ProcessExtensionsTests.cs:295]: Done in 41c71b9 — `VerifyAppliedAffinity` now lists missing processors over the full 64-bit width instead of the runtime `AffinityMaskWidth`, so the error message (and the bit-63 expectation in the test) is deterministic regardless of process bitness. Callers already constrain requested masks to the native width, so the wider listing bound is always correct. Thanks!

File: tests/Common.Tests/Diagnostics/ProcessExtensionsTests.cs [hunk 8]
```diff
@@ -38,18 +46,6 @@ public void SetSingleProcessorAffinity_should_throw_for_invalid_processor_number
...
-        // Act
-        var act = () => process.SetSingleProcessorAffinity(Environment.ProcessorCount);
-
-        // Assert
-        act.Should().Throw<ArgumentOutOfRangeException>();
-    }
-
     [Fact]
     public void SetSingleProcessorAffinity_should_throw_for_processor_number_beyond_native_mask_width()
     {
```
[the remark this answers is not in the document]
Comment (kploch) [reply tests/Common.Tests/Diagnostics/ProcessExtensionsTests.cs:51]: Done in 41c71b9 — added `GetEnabledProcessors_should_return_no_processors_for_a_zero_width_mask` asserting that an all-ones mask yields an empty result at width 0. Thanks!

File: tests/Common.Tests/Diagnostics/ProcessExtensionsTests.cs [hunk 9]
```diff
@@ -13,14 +13,22 @@ public void SetSingleProcessorAffinity_should_set_affinity_mask_for_valid_proces
...
-        enabledProcessors.Should().Contain(Environment.ProcessorCount - 1);
+            process.GetEnabledProcessors().Should().Equal(targetProcessor);
+        }
+        finally
+        {
+            KillAndReap(process);
+        }
     }
 
     [Theory]
```
[the remark this answers is not in the document]
Comment (kploch) [reply tests/Common.Tests/Diagnostics/ProcessExtensionsTests.cs:34]: Done in 41c71b9 — the acceptance theory gained `(1, 0)` and the throw theory gained `(1, 1)`, making the single-bit-mask boundary explicit. Thanks!

File: src/Common/Diagnostics/ProcessExtensions.cs [hunk 10]
```diff
@@ -139,39 +189,65 @@ public static IEnumerable<int> GetEnabledProcessors(this Process process)
...
+                                                  affinityMaskWidth,
+                                                  "The affinity-mask width must be between 1 and 64 bits.");
+        }
+
+        if (processorNumber < 0 || processorNumber >= affinityMaskWidth)
         {
             throw new ArgumentOutOfRangeException(paramName,
                                                   processorNumber,
-                                                  $"Processor number must be between 0 and {maxAddressable - 1} — the lesser of the machine's processor count ({processorCount}) and the native affinity-mask width ({affinityMaskWidth} bits).");
+                                                  $"Processor number must be between 0 and {affinityMaskWidth - 1} — the width of the native processor-affinity mask ({affinityMaskWidth} bits). Processor numbers for CPUs that do not exist on this machine are rejected by the operating system when the affinity is applied.");
```
[the remark this answers is not in the document]
Comment (kploch) [reply src/Common/Diagnostics/ProcessExtensions.cs]: Done in 47385f1 — the `ArgumentOutOfRangeException` message now describes only mask-width representability ("the exclusive upper bound is the width of the native processor-affinity mask"), and the helper's remarks defer platform behaviour to the applying methods (OS error, or interception via `VerifyAppliedAffinity` where the OS applies an intersection). Thanks!

File: tests/Common.Tests/Diagnostics/ProcessExtensionsTests.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (kploch) [reply tests/Common.Tests/Diagnostics/ProcessExtensionsTests.cs]: Done in 6c594b4 — the process is now declared with `using var`, consistent with the other child-process tests in the file, so the handle is released when the test completes. Thanks!
